Sources/Parsers/ColorsFileParser.swift

```
@@ -4,34 +4,70 @@
 import Foundation
-import Kanna
 import PathKit
 
-public protocol ColorsFileParser {
-  var colors: [String: UInt32] { get }
-}
-
 public enum ColorsParserError: Error, CustomStringConvertible {
   case invalidHexColor(string: String, key: String?)
   case invalidFile(reason: String)
+  case unsupportedFileType(supported: [String])
 
   public var description: String {
     switch self {
-    case .invalidHexColor(string: let string, key: let key):
+    case .invalidHexColor(let string, let key):
       let keyInfo = key.flatMap { " for key \"\($0)\"" } ?? ""
       return "error: Invalid hex color \"\(string)\" found\(keyInfo)."
-    case .invalidFile(reason: let reason):
+    case .invalidFile(let reason):
       return "error: Unable to parse file. \(reason)"
+    case .unsupportedFileType(let supported):
+      return "error: Unsupported file type. The supported file types are: \(supported.joined(separator: ", "))"
+    }
+  }
+}
+
+protocol ColorsFileTypeParser: class {
+  static var extensions: [String] { get }
+
+  init()
+  func parseFile(at path: Path) throws -> [String: UInt32]
+}
+
+public final class ColorsFileParser {
+  private var parsers = [String: ColorsFileTypeParser.Type]()
+  var colors = [String: UInt32]()
+
+  public init() {
+    register(parser: ColorsCLRFileParser.self)
+    register(parser: ColorsJSONFileParser.self)
+    register(parser: ColorsTextFileParser.self)
+    register(parser: ColorsXMLFileParser.self)
+  }
+
+  public func parseFile(at path: Path) throws {
+    guard let parserType = parsers[path.extension?.lowercased() ?? ""] else {
+      throw ColorsParserError.unsupportedFileType(supported: Array(parsers.keys))
+    }
+
+    let parser = parserType.init()
+    let colors = try parser.parseFile(at: path)
+
+    for (name, value) in colors {
+      self.colors[name] = value
+    }
+  }
+
+  func register(parser: ColorsFileTypeParser.Type) {
+    for ext in parser.extensions {
+      parsers[ext] = parser
```

Should we warn / error in case multiple parsers declare the same extension, or should it be standard and expected that the last registered one is the one with the higher priority on that duplicate extension?

~I'm also wondering if we'll always distinguish parsers only according to the file extensions (we could instead imagine a `static func canParseFile(file: Path) -> Bool` or something — so that the parser can also analyse the file structure to tell if it can parse the file, would allow multiple parsers to support JSON or XML extensions with different internal structure/format), but it's probably better to keep that feature for later, and only add it if we really need it one day, right? (YAGNI)~

---

We can throw an error for now, eventually we could have multiple parsers per extension (or mime-type) and should then define some priority rules, especially in a plugin scenario.

---

Yeah let's do that for now. We can still improve that later and make it more flexible in the future, step by step

---

I think that instead of throwing an error, when we have the warning system, this should trigger a warning. But leave the error in for now until we have it.

---

Sure. The warning closure could be part of the PR creating the unifying parser protocol (#39) as the closure will likely be part of this protocol anyway